Supervisor for KafkaIndexTask

dclim opened the pull request:
Manages the creation and monitoring of KafkaIndexTasks based on a SupervisorSpec submitted to /druid/indexer/v1/supervisor. See #1642 for realtime ingestion improvements overview. Closes #2635.

fjy commented:
@dclim this fails UT

dclim commented:
Should be fixed now, thanks!

dclim commented:
Here are more details on what is included and not included in this PR:

Kafka supervisor manages the creating and monitoring of Kafka indexing tasks, which are tasks that ingest events from Kafka using offset ranges instead of time intervals (see #2220). A supervisor is created by providing a supervisor spec which includes fields for (among other things):
  - the Kafka topic to ingest
  - the number of indexing tasks to spawn
  - the number of events each task should handle before persisting a segment and completing
  - the number of indexing task replicas
  - an ingestion spec and configuration that will be passed down to the created Kafka indexing tasks

The supervisor will then periodically run the following sequence of steps:
  - get the current number of partitions for the topic from Kafka
  - get the current running tasks for the data source we are supervising
  - using this information, make sure that we have the expected number of tasks processing the Kafka partitions

As an example, say we have a Kafka topic with 6 partitions, we want to split these partitions across 2 tasks, with each task handling 3000 events, and we want a replication factor of 2. This will create 4 tasks:
  - Task 1: (partitions 0, 2, 4), offsets (P0: 0-1000, P2: 0-1000, P4: 0-1000) - Replica 1
  - Task 2: (partitions 1, 3, 5), offsets (P1: 0-1000, P3: 0-1000, P5: 0-1000) - Replica 1
  - Task 3: (partitions 0, 2, 4), offsets (P0: 0-1000, P2: 0-1000, P4: 0-1000) - Replica 2
  - Task 4: (partitions 1, 3, 5), offsets (P1: 0-1000, P3: 0-1000, P5: 0-1000) - Replica 2

(Replica tasks are created with the same availability group so that they will be executed on different nodes.)

If:
  - Task 1 dies: the supervisor will create another task with the same parameters to maintain the desired replication factor: 
    - [Task 5: (partitions 0, 2, 4), offsets (P0: 0-1000, P2: 0-1000, P4: 0-1000)]

  - Task 1 succeeds: the supervisor will terminate redundant replicas after the segment has been published and will spawn tasks to handle the next range of offsets:
    - kill Task 3
    - create [Task 5: (partitions 0, 2, 4), offsets (P0: 1000-2000, P2: 1000-2000, P4: 1000-2000) - Replica 1]
    - create [Task 6: (partitions 0, 2, 4), offsets (P0: 1000-2000, P2: 1000-2000, P4: 1000-2000) - Replica 1]

As of now, this implementation is fully functional but has a number of areas that can be refined to be more efficient / a better user experience (numbered to facilitate discussion):

1) Currently, the supervisor waits for the previous set of tasks to complete before submitting the next set of tasks to handle the next offset range. This results in a pause in processing due to the time it takes to build/push/load the segment. No data is lost (assuming Kafka buffers all the messages during that time) but ingestion is bursty. We should be preemptively adding tasks to handle the next set of offsets as the previous set near the end of their range so that the new tasks begin handling data while the previous set are persisting. This requires being able to talk to the peons the tasks are running on and is enabled by #2419 .

2) Currently, when a task completes (the segment is published and has been loaded by a historical), the supervisor will kill any replica tasks that are doing the same thing since it is now redundant. This generates a failed status, where it would be better to have it either show a successful or stopped status.

3) Currently, the supervisor itself does not persist in any way and will die if the overlord crashes, restarts, or loses leadership in any other way. The supervisor spec should be written to metadata storage and read on overlord startup so that it can survive failures and cluster upgrades without the user needing to resubmit it every time. There already is an endpoint that can be used to stop running supervisors.

4) Currently, all replicas are given an identical set of partitions/offset ranges to process and then independently do their thing. This produces correct final output, but could potentially lead to realtime queries returning different results, depending on which worker is queried by the broker and if that worker is unable to keep up with the data being pushed into Kafka and is lagging behind. A solution would be to implement a leader/follower mechanism where the followers would check with the leader to determine what offsets they should be reading which would improve synchronization between tasks.

5) Currently there is no graceful shutdown mechanism for Kafka indexing tasks, which means that the supervisor is unable to provide any way of early stopping of spawned tasks. When the supervisor is stopped, it leaves the tasks running and they will continue to run until they have ingested up to the ending offset for all the partitions they are reading from. Early stopping (where the supervisor can tell the running tasks to stop ingesting, persist what they currently have, and then end) would make things like rolling updates somewhat nicer.

6) Currently, when a new supervisor starts up, it gets a list of running tasks and kills any Kafka indexing task that is building segments for its data source that it did not create (it actually does this every time the main loop runs). This is to prevent situations where tasks created by a previous supervisor are running but have a different ingestion spec that will lead to offsets from Kafka being ingested twice (as the segment's sequenceName is based on the partitions/offsets being read, so a task reading different partitions will generate a different sequenceName which will generate a new segment containing duplicated data). It would be better if the supervisor inspected the Kafka indexing task to see if it was processing the same partitions/offsets and allow it to finish if it is safe, even if it wasn't the one who created the task.

dclim commented:
@himanshug @gianm as discussed during the dev sync, I've added some details into what is included in this PR and what enhancements could be added. Let me know your thoughts.

nishantmonu51 commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -0,0 +1,94 @@
+
+  private final Map<String, Pair<Supervisor, SupervisorSpec>> supervisors = new HashMap<>();
+
+  public Set<String> getSupervisors()
```
rename to getSupervisorIDs ?

dclim replied:
sounds good, thanks!

nishantmonu51 commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -0,0 +1,94 @@
+    return supervisors.containsKey(id);
+  }
+
+  public String createSupervisor(SupervisorSpec spec)
```
Suggestions - 
- rename to startSupervisor() or createAndStartSupervisor() to be more explict ?
- possibly return boolean indicating whether a there existed a supervisor earlier or a new one is created now.

dclim replied:
sounds good, thanks!

nishantmonu51 commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -0,0 +1,94 @@
+  public void stopSupervisor(String id)
+  {
+    Pair<Supervisor, SupervisorSpec> pair = supervisors.get(id);
+    if (pair != null && pair.lhs != null) {
```
in which case pair.lhs can be null ? 
should we remove it from the map in if pair.lhs is null?

dclim replied:
@nishantmonu51 currently the only way pair.lhs could be null is if someone wrote a botched implementation of SupervisorSpec that returned null for SupervisorSpec::createSupervisor(), although this would have already NPE'd here when we tried to start it after creating it. I think it's fine to remove this check

gianm replied:
could also do a `Preconditions.checkNotNull(x, "x")` on the things we expect to be not-null, so if anyone *does* botch them, it's at least easy to debug.

dclim commented:
Added improvement 6: supervisor no longer kills all tasks on startup, but checks to see if the task matches the supervisor's spec and expected offsets and starts tracking it if it conforms. If it doesn't match, it will kill the task to prevent ingesting duplicate events.

himanshug commented:
@dclim nice write up. here are the things, i think, we need to have in the order of priority.

I believe number (3) is a MUST have . Supervisor must persist its state in metadata store because overlord going down or changing leadership is a very common case, even in case of rolling deployment.

(5) about  early stopping of tasks and related "user supplies number of events a task will handle", it is very difficult for users to figure out that number and many times it could change depending on hour of the day, day of the week. This might impact rolling deployment also where some tasks are never finishing.

(2) about remaining replica's showing FAILED status is confusing to users. It will require users to very carefully look at the overlord console task statuses to say whether ingestion is working fine or if there are genuine failures.

pjain1 commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,608 @@
+      log.debug("Task group [%d] pre-pruning: %s", groupId, taskGroup.taskStatuses);
+      String successfulTaskId = null;
+      Iterator<Map.Entry<String, Optional<TaskStatus>>> it = taskGroup.taskStatuses.entrySet().iterator();
+      while (it.hasNext()) {
```
Why not remove the TaskGroup for which all tasks were killed here from the `taskGroups` map otherwise new tasks will be again spawned for this TaskGroup at Line 397 ? Is that correct ? Am I missing something ?
Consider this scenario, if the taskCount is set to 3 and number of kafka partitions is 3, now i decreased the number of partitions form 3 to 1, this would cause all tasks to be killed in two TaskGroups.

dclim replied:
@pjain1 removing the TaskGroup when all tasks are killed would be okay; the only (minor) downside is that with the current logic the new retry tasks wouldn't be enqueued until the next cycle. However, I don't believe this is necessary since Kafka does not support decreasing the number of partitions, and I can't think of any other reason that the number of task groups would decrease.

pjain1 replied:
Also, does this PR handles the scenarios where one increases or decreases the `taskCount` and restarts the kafka supervisor ?

gianm replied:
There is only one situation where partitions can decrease- a kafka user deletes a topic and re-creates it with fewer partitions. But in that case it's not really the "same" topic (offsets are gonna get re-used too) so the supervisor is gonna get really confused and need to be re-set.

we could help people out in this situation with a "hard reset" option that stops the supervisor, and wipes the datasource metadata. that would let them resume from the start or end of the new topic, as they wish.

dclim replied:
@pjain1 yes, right now what would happen is the supervisor would come back up, discover the existing running tasks, and notice that they are processing a different allocation of partitions than what is specified in its spec (since the number of tasks have changed so the number of partitions processed per task also changed) - it would then forcibly kill these tasks and create new ones which would start reading from the last persisted offsets which are stored in the datasource metadata table. Assuming that Kafka's buffer is large enough that these events haven't been dropped, no data would be lost.

pjain1 replied:
@dclim great...I didn't knew that kafka does not allow decreasing the number of partitions in which case it should be fine

fjy commented:
@pjain1 any more comments?

gianm commented:
IMO the importance of the issues @dclim raised are, from most important to least,

- 3 (supervisor persisting) is critical
- 6 (avoiding killing "safe" tasks on startup) is important to make people feel like it's safe to restart overlords whenever they want (which is currently the case, and we want to keep it that way)
- 2 (unnecessary replicas being killed and FAILED) is important, as the current behavior will be confusing for users like @himanshug pointed out
- 1 (eliminating ingestion pauses) is going to be critical for users that have a real real-time requirement, and nice to have for other users
- 5 (early stopping) would be nice for getting schema updates out faster. Kafka tasks support restoreTasksOnRestart so I think this is not really needed for rolling updates (the tasks can restore after a MM restart)
- 4 (keeping replicas in sync) is nice but least important

Of those I think 3, 6, 2 are important to do before releasing the feature at all.

I think we can live without 1 for the first release but should do it in a follow-up.

I think 5, 4 are nice improvements but less critical than the others.

@himanshug - does this sound reasonable to you?

himanshug commented:
@gianm from https://github.com/druid-io/druid/pull/2656#issuecomment-197656072 , i believe that (6) is already done.
I agree with 3,6,2 being important.

(4) would be nice to have but very hard to do without adding complexity , not having it is fine. also, (1) is not too critical at least for now in the first cut.

currently user is required to set "the number of events each task should handle before persisting a segment and completing" . it will be very convenient if this is dynamically figured out by the system automatically. I thought (5) could probably help support this too in addition to speeding up schema updates as you pointed.

pjain1 commented:
@fjy I :+1: on @gianm suggestion apart from that I do not have any more comments

dclim commented:
@himanshug thanks for your feedback - what were you thinking of as an alternative to having the user specify the number of events a task will handle? It might be interesting if we allowed the user to specify a target for how long they want the task to run and then run an initial task to ingest a small number of events to get a feel of the events/sec and then use this to tune our offset ranges. Also like you said, if we had graceful shutdown we could guarantee tasks live as long as we want them to. Is that along the lines of what you were thinking? (have the user set the lifetime of the task in time vs number of events)

himanshug commented:
@dclim as discussed on dev sync it will be nice if supervisor starts the tasks with some very large arbitrary end-offset and then forces the push if/when segments reach some size threshold or some time (e.g. 30 mins) has elasped.

dclim commented:
Update:
The Kafka supervisor has been reworked and includes the following refinements:

- Task lifetime is now based on time rather than a number of ingested events - as @himanshug pointed out, this is easier for users to reason about and better handles fluctuating ingestion rates (or even situations where no events are ingested at all in a time period)
- #1 from above (subsequent tasks are started immediately instead of waiting until the previous set has finished publishing)
- #2 from above (tasks stopped by the supervisor report a SUCCESS status with log messages in the task showing that it was asked to stop early)
- #3 from above (supervisor specs are persisted in metadata storage and will automatically load and begin running when the overlord starts (or takes leadership)
- #5 from above (KafkaIndexTasks can be stopped early and instructed to publish - telling the supervisor to shutdown via the HTTP API will cause a controlled (synchronized across replicas) early publish and shutdown of the managed tasks)
- #6 from above (on startup, supervisor now inspects and adopts already-running tasks that match the supervisor's spec and expected start offset)

The supervisor logic has changed somewhat from the above description to accommodate time-based task lifetimes and no pauses between tasks. I'll post an updated overview of the mechanism, but as is this PR should be ready for review.

dclim commented:
Some additional notes regarding the time-based task lifetime design:

Previously, the supervisor created KafkaIndexTasks, provided a starting and ending offset, and when the task completed, the supervisor created another task (to handle the next set of offsets if the task succeeded or to reprocess the same range of offsets if the task failed). With the updated design, the user provides a `taskDuration` value in the supervisor spec which is used to manage the lifetime of tasks based on time rather than starting and ending offsets. The high-level mechanism works as follows:

1. When the supervisor starts up, assuming no tasks are already running, it will create up to `taskCount` * `replicas` number of tasks. Each of these tasks will be provided with a start offset per partition corresponding to the last offset that was read and published in a segment (from the dataSource metadata table) or if this doesn't exist, either the earliest or latest offset from Kafka depending on configuration. The **end offset** is now always set to Long.MAX_VALUE. This is because we can't know how many events we will be reading during `taskDuration` so we initially set the tasks to read indefinitely.
2. Once the task is accepted by a worker and is started, the supervisor queries the task for its starting time and stores this information. It schedules an event to fire at `startTime` + `taskDuration`.
3. When the event is fired, the supervisor then executes the logic to instruct the tasks to end reading and begin publishing their segments in a coordinated way. Coordination is important, because if we just told the tasks to stop and publish, replica tasks which are expected to have read the same data and publish exactly the same segments are not guaranteed to produce identical segments since the Kafka consumers might be at different offsets for different partitions when they receive the stop command. To coordinate the tasks, we do the following:
    
  a. Tell all tasks handling a particular set of partitions (in the code referred to as a task group) to pause and return the offsets it had read up to before it was paused.

  b. Build a list of the highest offset for each partition read by any of the tasks.

  c. Tell all of the tasks to change their ending offset from Long.MAX_VALUE to the list generated in step b. Using the highest offset (and not for example the highest offset + 1) guarantees that all the data is available in Kafka (since at least one of the tasks has already read it) and that the tasks will always be able to complete.

  d. Tell all of the tasks to resume reading.

  The tasks will then all read to the same ending offset for every partition and begin publishing their segment.
4. The supervisor will then immediately create tasks to continue reading from Kafka starting from the ending offsets determined in step 3. This will allow for continuous ingestion from Kafka (with a few seconds of delay as the new task is accepted by a worker and starts up). At the same time, the previous tasks which are now publishing are moved to a list of 'pending completion' tasks and assigned a timeout specified by the `completionTimeout` config.
5. If a task which is supposed to be publishing does not finish before the timeout, the task is deemed to have failed and is terminated. Furthermore, segments generated by KafkaIndexTasks must contain contiguous offsets, so a failure to publish by a previous task means that the next tasks that were reading from the same partitions are now invalid. The supervisor handles this by killing the tasks reading those partitions and then re-creating new tasks that start from the last successfully published offset (tracked in the dataSource table).

pjain1 commented:
Does the `completionTimeout` should always be less than `taskDuration` ? otherwise in a situation where are two sets of pending tasks the older set may fail but the newer set may publish. I guess it may not happen because of metadata equal checks in `updateDataSourceMetadataWithHandle` method ?

dclim commented:
@pjain1 good question - there shouldn't need to be any condition on completionTimeout < taskDuration. If there are two sets of pending tasks (plus the currently reading task set) and the older set fails, it will kill itself, the newer pending tasks, as well as the currently reading set since all of them will now be producing invalid (non-contiguous) segments. The supervisor will then create new tasks that'll start from the offsets in the last successful segment (which would be the starting offsets of that older task set that failed).

In general, you wouldn't want time-to-publish >> taskDuration, since that would mean that you'd be spawning new tasks faster than they're completing and your workers will eventually run out of capacity, but in terms of correctness, there shouldn't be any issues.

himanshug commented:
@dclim can you add some user documentation ?

dclim commented:
Yes, will write some up.

dclim commented:
@himanshug docs added: https://github.com/dclim/druid/blob/kafka-supervisor/docs/content/ingestion/kafka-ingestion.md

schmee commented:
@dclim I read through the docs and noticed that the `ioConfig` takes a `topic`, unlike Tranquility that takes a `topicPattern`. Is this a fundamental limitation of the new KafkaIndexTask? If so, what is the best way to emulate how Tranquility handles topics?

dclim commented:
@schmee yeah, the KafkaIndexTask was written to read a single topic only; I don't believe that it's a fundamental limitation, but some work would be required to support topic patterns. We use topic patterns in our Druid cluster as well so I definitely think there's value in supporting them.

What is your use case for topicPatterns? Are you reading from multiple topics into one dataSource or multiple topics into multiple dataSources (using the same ingestion spec)?

schmee commented:
@dclim Our use-case would be multiple topic single datasource. Our Kafka topics are structured according to `<site>/<data_type>`, and we would like to ingest multiple entities and data types in the same datasource.  Without topic patterns, if we want to combine 20 entities and 3 data types, and have 2x replication, that would amount to 120 tasks. This is a major concern for performance as well as usability, since we would have to join all the datasources client-side. Is there any way to support this use-case that doesn't require us to restructure our Kafka topics?

gianm commented:
@schmee For now, the best approach would be to use something like Samza or Kafka Streams to combine your topics into a new, single topic per Druid dataSource. This job could also do some other pre-processing, if necessary.

The KafkaIndexTask could in principle do multiple topics in the future, but we didn't implement that for the initial release.

himanshug commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java:
```diff
@@ -0,0 +1,326 @@
+        }
+
+        int responseCode = response.getStatus().getCode();
+        if (responseCode / 100 == 2) {
```
in what case anything but 200 is acceptable?

himanshug replied:
ok, i see that sometimes index task responds with ACCEPTED .

himanshug commented on server/src/main/java/io/druid/metadata/SQLMetadataSupervisorManager.java:
```diff
@@ -0,0 +1,250 @@
+  {
+    return dbTables.get().getSupervisorTable();
+  }
+}
```
can you please add tests for this class? some of the queries used here are non-trivial.

dclim replied:
Sounds good

himanshug commented on server/src/main/java/io/druid/metadata/SQLMetadataSupervisorManager.java:
```diff
@@ -0,0 +1,250 @@
+          @Override
+          public Void withHandle(Handle handle) throws Exception
+          {
+            final String version = new DateTime().toString();
```
not sure why we need to maintain versioned history of a given supervisor's spec?

dclim replied:
@gianm felt it would be good to keep a history, Gian do you have any thoughts on this?

gianm replied:
I think history is pretty useful here, it'll help us track down which changes were made and when. Sort of like how we also keep a history of changes to the coordinator rules table.

himanshug replied:
ok, does it make sense to limit the number of items in the history?

gianm replied:
We don't limit history for any other kind of metadata (rules, segments, tasks) so I don't think we should limit history here. users could delete old rows if they really want to, or it might also make sense to introduce a retention option for more kinds of metadata in a future PR.

himanshug commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorResource.java:
```diff
@@ -0,0 +1,230 @@
+      return f.apply(supervisorManager.get());
+    } else {
+      // Encourage client to try again soon, when we'll likely have a redirect set up
+      return Response.status(Response.Status.SERVICE_UNAVAILABLE).build();
```
yes, there should be redirection in place already, this is really a SERVER_ERROR scenario, request shouldn't land here in general if this overlord is not leader.

dclim replied:
I borrowed this method from OverlordResource, I assumed that there might be some kind of timing issue between when we lose leadership and when the servlet starts returning redirect responses which is why the code is here, is this not actually an issue anymore?

himanshug replied:
you are doing the right thing. i was just confirming, the timing issue will still be there.

himanshug commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorResource.java:
```diff
@@ -0,0 +1,230 @@
+          public Response apply(SupervisorManager manager)
+          {
+            if (manager.hasSupervisor(spec.getId())) {
+              return Response.status(Response.Status.CONFLICT)
```
ok, so if user needs to make some change to spec e.g. changing the schema, then they are supposed to first shutdown the supervisor and then do the POST.
this one should be in the doc i believe.can you please add a section in the doc about how to make changes, also if user is supposed to do anything special as part of roling deployment.

dclim replied:
I see you found the section in the doc, but yes I think it would be good to make it more of an explicit section with a header, I can do that.

himanshug commented on docs/content/ingestion/kafka-ingestion.md:
```diff
@@ -0,0 +1,234 @@
+managed tasks to stop reading and begin publishing their segments immediately. The call to the shutdown endpoint will
+return after all tasks have been signalled to stop but before the tasks finish publishing their segments. Schema
+rollovers are managed by first shutting down the supervisor, waiting for the running tasks to complete, and then
+submitting the new supervisor spec.
```
nice doc

himanshug commented on common/src/main/java/io/druid/metadata/MetadataStorageTablesConfig.java:
```diff
@@ -83,7 +86,8 @@ public MetadataStorageTablesConfig(
       @JsonProperty("taskLock") String taskLockTable,
-      @JsonProperty("audit") String auditTable
+      @JsonProperty("audit") String auditTable,
+      @JsonProperty("supervisors") String supervisorTable
```
needs doc update

dclim replied:
Sounds good

himanshug commented:
@gianm @dclim @schmee i agree about keeping scope of this PR to single topic support, it would be nice to have that in future.. i would let this impl get settled a little bit by releasing and running on some production clusters and iron out [if any] stability issues first.

himanshug commented on server/src/main/java/io/druid/metadata/SQLMetadataSupervisorManager.java:
```diff
@@ -0,0 +1,250 @@
+                        + "FROM %1$s r "
+                        + "INNER JOIN(SELECT spec_id, max(version) as version FROM %1$s GROUP BY spec_id) latest "
+                        + "ON r.spec_id = latest.spec_id and r.version = latest.version",
+                        getSupervisorsTable()
```
limiting items in history may potentially help the performance of this query too.

himanshug commented:
@dclim 
:+1: besides https://github.com/druid-io/druid/pull/2656#discussion_r61903346 , i know it does not matter if history stays small in size but will leave it up to you.

gianm commented on docs/content/ingestion/kafka-ingestion.md:
```diff
@@ -0,0 +1,240 @@
+
+# Kafka Ingestion
+
+The recommended way of ingesting data from Kafka is to use the `kafka-indexing-service` core extension (see
```
I like the optimism but this language is a bit strong for first release :)

The language here should be telling people that this is an experimental feature, API subject to change, etc.

dclim replied:
haha, sounds good

gianm commented on docs/content/ingestion/kafka-ingestion.md:
```diff
@@ -0,0 +1,240 @@
+|`maxRowsInMemory`|Integer|The number of rows to aggregate before persisting. This number is the post-aggregation rows, so it is not equivalent to the number of input events, but the number of aggregated rows that those events result in. This is used to manage the required JVM heap size. Maximum heap memory usage for indexing scales with maxRowsInMemory * (2 + maxPendingPersists).|no (default ==  ...
+|`maxRowsPerSegment`|Integer|The number of rows to aggregate into a segment; this number is post-aggregation rows.|no (default == 5000000)|
+|`intermediatePersistPeriod`|ISO8601 Period|The period that determines the rate at which intermediate persists occur.|no (default == PT10M)|
+|`basePersistDirectory`|String|The local directory used for persisted data.|no (default == Java temp directory)|
```
`basePersistDirectory` is ignored by tasks (it's overridden to the task dir) so we can take this out.

gianm commented on docs/content/ingestion/kafka-ingestion.md:
```diff
@@ -0,0 +1,240 @@
+|`replicas`|Integer|The number of replica sets, where 1 means a single set of tasks (no replication). Replica tasks will always be assigned to different workers to provide resiliency against node failure.|no (default == 1)|
+|`taskCount`|Integer|The maximum number of *reading* tasks in a *replica set*. This means that the maximum number of reading tasks will be `taskCount * replicas` and the total number of tasks (*reading* + *publishing*) will be higher than this. See 'Capacity Planning' below for more details. The number of reading tasks will be less than `taskCount` if `taskCount > {numKafkaPartitions}`.|no (defau ...
+|`taskDuration`|ISO8601 Period|The length of time before tasks stop reading and begin publishing their segment. Note that segments are only pushed to deep storage and loadable by historical nodes when the indexing task completes.|no (default == PT1H)|
+|`consumerProperties`|Map<String, String>|A map of properties to be passed to the Kafka consumer. Note that 'bootstrap.servers' will be overridden with the value in `kafkaBrokers`.|no|
```
[minor] I think it would more user friendly to get `bootstrap.servers` from here rather than have a separate `kafkaBrokers` thing. (it better leverages users' existing knowledge of kafka and existing kafka consumer configs)

dclim replied:
sounds good

gianm commented on docs/content/ingestion/kafka-ingestion.md:
```diff
@@ -0,0 +1,240 @@
+
+Following from the previous section, schema and configuration changes are managed by first shutting down the supervisor
+with a call to the `POST /druid/indexer/v1/supervisor/<supervisorId>/shutdown` endpoint, waiting for the running tasks
+to complete, and then submitting the updated schema via the `POST /druid/indexer/v1/supervisor` create supervisor
```
what happens if you submit a new supervisor before the old tasks exit? (users might want to read about this in the docs)

dclim replied:
Will add it to the docs. If the new supervisor has a different spec, it will kill the previous tasks, which may or may not have already published the segment it was working on.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskModule.java:
```diff
@@ -39,7 +40,8 @@
                 new NamedType(KafkaIOConfig.class, "kafka"),
-                new NamedType(KafkaTuningConfig.class, "kafka")
+                new NamedType(KafkaTuningConfig.class, "kafka"),
+                new NamedType(KafkaSupervisorSpec.class, "kafka")
```
kafka kafka kafka kafka kafka kafka

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,1263 @@
+    // this task group has completed successfully, at which point this will be destroyed and a new task group will be
+    // created with new starting offsets. This allows us to create replacement tasks for failed tasks that process the
+    // same offsets, even if the values in [partitionGroups] has been changed.
+    Map<Integer, Long> partitionOffsets = new HashMap<>();
```
It looks like `partitionOffsets` and `tasks` aren't being reassigned, so please make them final.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,1263 @@
+  private final Timer timer = new Timer(true);
+  private final Object stopLock = new Object();
+
+  private KafkaConsumer consumer;
```
This should be volatile since it's initialized in the thread calling `start` and then used in a different thread after that.

dclim replied:
thanks!

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,1263 @@
+    }
+
+    firstRunTime = DateTime.now().plus(ioConfig.getStartDelay());
+    timer.scheduleAtFixedRate(
```
most other Druid code uses `Execs.scheduledSingleThreaded(threadName)` rather than a timer. That makes sure we always get a daemon thread with a nice name. Looks like you already have it using a daemon thread, but a name would be nice too.

So I would either switch this to a scheduled executor (with a nice name) or use the Timer constructor that takes a thread name.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,1263 @@
+      }
+      lastRunTime = nowTime;
+
+      runInternal();
```
This should log something at INFO level when it decides to actually run.

gianm replied:
(assuming that's not being logged somewhere else – didn't see anything like that yet)

dclim replied:
@gianm ah, I actually had it info log a summary of the tasks it was managing but then upon looking at the logs thought it was getting too spammy so I removed it yesterday. The log messages looked like this:
```
2016-05-02T22:01:34,527 INFO [KafkaSupervisor-github-v2] io.druid.indexing.kafka.supervisor.KafkaSupervisor - {id='github-v2', generationTime=2016-05-02T22:01:34.527Z, payload={dataSource='github-v2', topic='github', partitions=3, replicas=1, durationSeconds=3600, active=[{id='index_kafka_github-v2_93416522dbb8099_olhpaebd', startTime=2016-05-02T21:33:28.394Z, remainingSeconds=1913}, {id='index_kafka_github-v2_965719e6f2b23f1_pkeljcng', startTime=2016-05-02T21:32:45.640Z, remainingSeconds=1871}], publishing=[]}}
2016-05-02T22:01:34,548 INFO [KafkaSupervisor-twitter-v2] io.druid.indexing.kafka.supervisor.KafkaSupervisor - {id='twitter-v2', generationTime=2016-05-02T22:01:34.548Z, payload={dataSource='twitter-v2', topic='twitter', partitions=3, replicas=2, durationSeconds=3600, active=[{id='index_kafka_twitter-v2_1c4fab566153c68_kbpjknfe', startTime=2016-05-02T21:32:28.397Z, remainingSeconds=1853}, {id='index_kafka_twitter-v2_1c4fab566153c68_diiajjaa', startTime=2016-05-02T21:32:31.193Z, remainingSeconds=1856}, {id='index_kafka_twitter-v2_53f8a23dfcdfa3a_mhofcgeb', startTime=2016-05-02T21:26:24.668Z, remainingSeconds=1490}, {id='index_kafka_twitter-v2_53f8a23dfcdfa3a_dblegedd', startTime=2016-05-02T21:26:22.060Z, remainingSeconds=1487}], publishing=[]}}
2016-05-02T22:01:54,527 INFO [KafkaSupervisor-github-v2] io.druid.indexing.kafka.supervisor.KafkaSupervisor - {id='github-v2', generationTime=2016-05-02T22:01:54.527Z, payload={dataSource='github-v2', topic='github', partitions=3, replicas=1, durationSeconds=3600, active=[{id='index_kafka_github-v2_93416522dbb8099_olhpaebd', startTime=2016-05-02T21:33:28.394Z, remainingSeconds=1893}, {id='index_kafka_github-v2_965719e6f2b23f1_pkeljcng', startTime=2016-05-02T21:32:45.640Z, remainingSeconds=1851}], publishing=[]}}
```
What do you think?

gianm replied:
Hmm, I think ideally what I'd like to see at INFO is,

1) something on startup saying, hey, we started!
2) updates whenever something changes (when kafka partitions are added, when tasks are created or destroyed or paused or whatever, stuff like that)
3) periodic messages when `runInternal` runs. Maybe not every single run if it happens really often (like more than a couple times a minute is probably excessive) but enough such that I know from the logs that the supervisor is still waking up and not stuck. As a user I'd be looking for that if I expect it to be doing something that it's not doing – the first thing to check is whether it's wedged

dclim replied:
Okay that sounds good. I had logs for all 3 but removed the last one thinking it might be too verbose for info level and moved it to debug, will fix.

gianm commented on server/src/main/java/io/druid/indexing/overlord/supervisor/Supervisor.java:
```diff
@@ -0,0 +1,28 @@
+public interface Supervisor
+{
+  void start();
+  void stop(boolean stopGracefully);
```
Could you please document what `stopGracefully` means?

gianm commented on indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java:
```diff
@@ -1118,6 +1141,7 @@ private void taskComplete(
 
     // Notify interested parties
+    TaskRunnerUtils.notifyStatusChanged(listeners, taskStatus.getId(), taskStatus);
     taskRunnerWorkItem.setResult(taskStatus);
```
Please switch these to keep the setResult -> notifyStatusChanged order consistent every time we do that. I think this matters because setResult is what updates TaskStorage (through a callback) and notifyStatusChanged is what triggers another supervisor run. Since supervisor does a getStatus we want that update to have already happened.

dclim replied:
Makes sense, nice catch.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,1266 @@
+      // if this task has run longer than the configured duration, signal all tasks in the group to persist
+      if (earliestTaskStart.plus(ioConfig.getTaskDuration()).isBeforeNow()) {
+        log.info("Task group [%d] has run for [%s]", groupId, ioConfig.getTaskDuration());
+        Map<Integer, Long> endOffsets = signalTasksToFinish(groupId);
```
What happens if the supervisor crashes while this code is running? Will the next supervisor that starts up to take its place be able to figure out what's going on?

gianm replied:
Or if a supervisor crashes, in general, while a task group is finishing or partially finishing,

- maybe it had asked half of them to persist but then crashed before the others were asked to persist?
- maybe it had asked all of them to persist, and they're pending but haven't finished yet

dclim replied:
The trigger for a task being "done" is based on the startTime reported by the task itself (it's "cached" in the supervisor's state but is strictly based on data from the task), so if the supervisor is in the middle of processing tasks that should be finished and then crashes, when the supervisor restarts and newly discovers the currently running tasks, if the tasks match the supervisor's spec and starting offset it'll start tracking the task, then it'll grab the start time from the task and then immediately realize that the task has run for the expected duration and will signal the tasks to finish or whatever it has to do.

If the supervisor crashed after issuing a pause to the task but before issuing a set-end-offsets-and-resume, the next supervisor would try to pause it again (which would succeed) and then continue to calculate the end offsets and start the publishing.

If it did get as far as starting the publishing and then crashed, the task state would have switched from READING to PUBLISHING and then when the next supervisor starts, when discoverTasks() runs and picks up those tasks, it'll check if it's in PUBLISHING state and if so it'll add it to the pendingCompletionTaskGroups of the new supervisor and track it as a publishing task.

dclim commented:
Fixed an issue where if the supervisor crashed after signalling a task to begin publishing but before creating the next task, the succeeding supervisor would create the new task with the same starting offsets as the publishing task. It will now create the new task starting from where the publishing task ended. Also added a test for this.

gianm commented:
```
Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 4.086 sec <<< FAILURE! - in io.druid.server.coordinator.DruidCoordinatorTest
testCoordinatorRun(io.druid.server.coordinator.DruidCoordinatorTest)  Time elapsed: 1.918 sec  <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:555)
	at org.junit.Assert.assertEquals(Assert.java:542)
	at io.druid.server.coordinator.DruidCoordinatorTest.testCoordinatorRun(DruidCoordinatorTest.java:374)
```

gianm commented:
@dclim looking good!

👍
